File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.rb
# @param {Integer[]} prices
# @return {Integer}
def max_profit(prices)
    n = prices.length 

    min_price = Float::INFINITY
    max_profit = 0 

    for price in prices do 
        min_price = [price, min_price].min 
        profit = price - min_price 
        max_profit = [max_profit, profit].max 
    end 

    return max_profit
end

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)
        min_price = float('inf')
        mp = 0 
        
        for i in range(n):
            min_price = min(prices[i], min_price)
            profit = prices[i] - min_price 
            mp = max(mp, profit)

        return mp 
        
        